dspinner/pin_test.go

```diff
@@ -0,0 +1,1056 @@
+	var err error
+
+	for i := range nodes {
+		err = p.Pin(ctx, nodes[i], recursive)
+		if err != nil {
+			panic(err)
+		}
+	}
+	err = p.Flush(ctx)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func unpinNodes(nodes []ipld.Node, p ipfspin.Pinner) {
+	ctx := context.Background()
+	var err error
+
+	for i := range nodes {
+		err = p.Unpin(ctx, nodes[i].Cid(), true)
+		if err != nil {
+			panic(err)
+		}
+	}
+	err = p.Flush(ctx)
+	if err != nil {
+		panic(err)
+	}
+}
+
+type batchWrap struct {
+	ds.Datastore
+}
+
+func (d *batchWrap) Batch() (ds.Batch, error) {
+	return ds.NewBasicBatch(d), nil
+}
+
+func makeStore() (ds.Datastore, ipld.DAGService) {
+	ldstore, err := lds.NewDatastore("", nil)
+	if err != nil {
+		panic(err)
+	}
+	var dstore ds.Batching
+	dstore = &batchWrap{ldstore}
+
+	bstore := blockstore.NewBlockstore(dstore)
+	bserv := bs.New(bstore, offline.Exchange(bstore))
+	dserv := mdag.NewDAGService(bserv)
+	return dstore, dserv
+}
+
+// BenchmarkLoadRebuild loads a pinner that has some number of saved pins, and
+// compares the load time when rebuilding indexes to loading without rebuilding
+// indexes.
+func BenchmarkLoadRebuild(b *testing.B) {
+	dstore, dserv := makeStore()
+	pinner := New(dstore, dserv)
+
+	nodes := makeNodes(4096, dserv)
+	pinNodes(nodes, pinner, true)
+
+	b.Run("RebuildTrue", func(b *testing.B) {
+		for i := 0; i < b.N; i++ {
+			dstore.Put(dirtyKey, []byte{1})
+
+			_, err := LoadPinner(dstore, dserv)
+			if err != nil {
+				panic(err.Error())
+			}
+		}
+	})
+
+	b.Run("RebuildFalse", func(b *testing.B) {
+		for i := 0; i < b.N; i++ {
+			dstore.Put(dirtyKey, []byte{0})
+
+			_, err := LoadPinner(dstore, dserv)
+			if err != nil {
+				panic(err.Error())
+			}
+		}
+	})
+}
+
+// BenchmarkNthPins shows the time it takes to create/save 1 pin when a number
+// of other pins already exist.  Each run in the series shows performance for
+// creating a pin in a larger number of existing pins.
+func BenchmarkNthPin(b *testing.B) {
+	dstore, dserv := makeStore()
+	pinner := New(dstore, dserv)
+	pinnerIPLD := ipldpinner.New(dstore, dserv, dserv)
+
+	for count := 1000; count <= 10000; count += 1000 {
+		b.Run(fmt.Sprint("PinDS-", count), func(b *testing.B) {
+			benchmarkNthPin(b, count, pinner, dserv)
+		})
+
+		b.Run(fmt.Sprint("PinIPLD-", count), func(b *testing.B) {
+			benchmarkNthPin(b, count, pinnerIPLD, dserv)
+		})
+	}
+}
+
+func benchmarkNthPin(b *testing.B, count int, pinner ipfspin.Pinner, dserv ipld.DAGService) {
+	ctx := context.Background()
+	nodes := makeNodes(count, dserv)
+	pinNodes(nodes[:count-1], pinner, true)
+	b.ResetTimer()
+
+	which := count - 1
+	for i := 0; i < b.N; i++ {
+		// Pin the Nth node and Flush
+		err := pinner.Pin(ctx, nodes[which], true)
+		if err != nil {
+			panic(err)
+		}
+		err = pinner.Flush(ctx)
+		if err != nil {
+			panic(err)
+		}
+		// Unpin the nodes so that it can pinned next iter.
+		b.StopTimer()
+		err = pinner.Unpin(ctx, nodes[which].Cid(), true)
```

Comment:
I specifically do not want to call `Flush` here since I only want to benchmark the time it takes to call `Pin` + `Flush` to simulate normal pinning.  The unpin is just so that the next iteration will work.

Reply:
The Unpin + potential Flush aren't being counted by the timer since it's between a stop and start, no?

In all likelihood this doesn't matter as long as the datastore is in memory, but if we switch the datastore to be on disk then we'd probably end up cheating and never actually writing to disk.

Reply:
Done.
Right, eliminated chance of datastore cheating; benchmark not affected due to timer stop/start.